pmacApp/src/pmacController.cpp

```diff
@@ -4566,34 +4680,104 @@ asynStatus pmacController::tScanCalculateVelocityArray(double *positions, double
         status = asynError;
         break;
       }
+      if(index > 0) {
+        if(index < numPoints-1){
+          // Get previous position from the current buffer
+          previousPos = positions[index-1];
+        } else {
+          // Next position and time not available, raise the flag for calculating the velocity at next iteration
+          tScanPendingPoint_ |= (1 << axis) ;
+          break;
+        }
+      } else if(pTrajectory_->getNoOfValidPoints()>1) {
+        // Get previous position from last buffer sent
+        previousPos = tScanPrevBufferPositions[axis][1];
+
+      } else {
+        // Get previous position from last demand Q7x
+        // Set previous velocity as zero
+        sprintf(command, "&%dQ%d", csNum, (71+axis));
+        if (pBroker_->immediateWriteRead(command, reply) == asynSuccess) {
+          sscanf(reply,"%lf",&previousPos);
+          previousVel = 0.0;
+        } else {
+          debug(DEBUG_ERROR, functionName, "Failed to send command", command);
+          status = asynError;
+          break;
```

Comment:
This code block is used twice in this section.  Maybe worth turning it into a function

Reply:
Done in [f884372](https://github.com/DiamondLightSource/pmac/pull/124/commits/f884372fd8fc88e73593bfd202c8d77fc9989792)